Implemented core database tables

alvyn279 opened the pull request:
- Added necessary classes and relational structure to `models.py`
- Satisfy issue #35

alvyn279 commented:
Hey guys, please have a look at the tables implementation I came up with.

One important thing I did not follow from our diagram (shown in the comments of issue #35) was the idea of lists as Fields. Instead of having a list of tags for `models.Questions`, I thought it would be simpler to add a foreign key in `models.Tags` that would keep track of all the tags for a single given question.

On another note, models.Answers has a boolean field that indicates if the answer is the correct one chosen by the one who asked the question. This will allow easier _switching on/switching off_ of the correct answer for a question

Lercerss commented on SA3/qa_web/models.py:
```diff
@@ -3,12 +3,48 @@
+    A comment can be posted for an entire question or an answer
+    There can be many comments to questions and answers
+    """
+    question = models.ForeignKey(Questions, null=True, on_delete=models.CASCADE)
```
I believe this is the correct way to implement the relation between Comments and Posts. Unfortunately it seems django models have limitations in terms of inheritance therefore the `post = models.ForeignKey(Post, null=True, on_delete=models.CASCADE` is not possible.

alvyn279 commented on SA3/qa_web/models.py:
```diff
@@ -3,12 +3,48 @@
+    Parent meta class that describes a post as a publication on the website with certain content
+    created by a certain owner
+    This will be used for multi-table inheritance
+    """
```
Very convenient!

jetcanfly replied:
Good advice.

alvyn279 commented:
Hey guys, so I adjusted the models to follow an abstract parent class like @Lercerss suggested. This required me to wipe all the database (...). Fear not, I have recreated the necessary users and a question that indicates that the `forms.login`, `forms.questions` and `signup` still work fine. I deleted my first question which explains the **no.2** id for the single question present on the index page.

@ALL Please take a look, review and/or comment

alvyn279 commented:
It would be amazing to get this reviewed and merged to `master` by the end of the afternoon for all the other team members to make use of this in their features implementation.

@jetcanfly @Lercerss

jetcanfly commented on SA3/qa_web/models.py:
```diff
@@ -3,12 +3,52 @@
+    This will be used for inheritance
+    """
+    content = models.TextField(null=True)
+    owner = models.ForeignKey(User, null= True, on_delete=models.SET_NULL)
```
The Django document raises some concerns about 'reference User directly'. It's suggested that using it this way may be better:
owner = models.ForeignKey(
        settings.AUTH_USER_MODEL,
or use get_user_model()
As [details](https://docs.djangoproject.com/en/2.0/topics/auth/customizing/).

Lercerss replied:
Good thing you pointed this out, if we want to plan ahead for #32 where we will probably have to define a custom User, this would be the right thing to do.

alvyn279 replied:
Great observation, I did not consider that beforehand. First of all, since we are using only one app in our project, specifying the `AUTH_USER_MODEL` will be rather unnecessary, as `User` is already the _else_ condition for its use. 

![image](https://user-images.githubusercontent.com/29107916/36077496-a162c888-0f39-11e8-8e5d-ea9ffa2f08f6.png)

Furthermore, considering using versatile code for multiple apps should be defined at the start of the project.

![image](https://user-images.githubusercontent.com/29107916/36077475-5aae2d92-0f39-11e8-86fd-e912169e4b81.png)

jetcanfly replied:
Good comment. Let's leave it.

alvyn279 replied:
If we want the change the User profile in a further issue (#32), I suggest we reconsider this right away as doing so requires a correct initial migration. Explanation [here](https://docs.djangoproject.com/en/2.0/topics/auth/customizing/#changing-to-a-custom-user-model-mid-project)

jetcanfly replied:
[Here ](https://simpleisbetterthancomplex.com/tutorial/2016/07/22/how-to-extend-django-user-model.html)is an article about extending django User model. If we don't want to change authentication process in #32 , we could just use option 2 or option 4.

jetcanfly commented on SA3/qa_web/models.py:
```diff
@@ -3,12 +3,52 @@
+    The question (or thread) initiates the interaction between users' answers,
+    comments, and voting
+    """
+    title = models.CharField(max_length=300, null=True)
```
Also in [django document](https://docs.djangoproject.com/en/2.0/ref/models/fields/), it shows us how 'null' works as a field option.  It's suggested that 'Avoid using null on string-based fields such as CharField and TextField. If a string-based field has null=True, that means it has two possible values for “no data”: NULL, and the empty string.'

alvyn279 replied:
Yes, once again this is true. However, I think that keeping a field to `NULL` (if there was no instantiation of that field) is a good practice. In the case of these string-based field, it might be useful for distinguishing `""` and `NULL` (no data at all).

jetcanfly commented on SA3/qa_web/models.py:
```diff
@@ -3,12 +3,52 @@
+    question = models.ForeignKey(Questions, null=True, on_delete=models.CASCADE)
+    answer = models.ForeignKey(Answers, null=True, on_delete=models.CASCADE)
+    
+class Tags(models.Model):
```
I found a nice lib [here](https://github.com/alex/django-taggit) . It seems easy to manipulate. Maybe we could consider using his wheel. :P

alvyn279 commented:
The backend is now set for a more versatile user (expected for issue #32).

alvyn279 commented on SA3/qa_web/models.py:
```diff
@@ -1,14 +1,57 @@
 
+class User(AbstractUser):
+    pass
+    
```
This is where we would extend the user profile for #32.
